File: 744-network-delay-time/network-delay-time.java
class Solution {
    // https://leetcode.com/problems/network-delay-time/

    public int networkDelayTime(int[][] times, int n, int k) {
        // Build graph
        Map<Integer, List<Pair<Integer, Integer>>> graph = new HashMap<>();
        for (int[] time : times) {
            int u = time[0];
            int v = time[1];
            int wt = time[2];
            graph.computeIfAbsent(u, key -> new ArrayList<>()).add(new Pair<>(v, wt));
        }

        // 1-based indexing
        PriorityQueue<Pair<Integer, Integer>> pq = new PriorityQueue<>(Comparator.comparingInt(Pair::getKey));
        pq.offer(new Pair<>(0, k));

        int[] costings = new int[n + 1];
        Arrays.fill(costings, Integer.MAX_VALUE);
        costings[k] = 0;

        Set<Integer> visited = new HashSet<>();

        int maxDelay = 0;

        while (!pq.isEmpty()) {
            Pair<Integer, Integer> pair = pq.poll();
            int costSoFar = pair.getKey();
            int node = pair.getValue();

            if (visited.contains(node)) {
                continue;
            }

            visited.add(node);

            if (graph.containsKey(node)) {
                for (Pair<Integer, Integer> neighborPair : graph.get(node)) {
                    int nei = neighborPair.getKey();
                    int wt = neighborPair.getValue();
                    if (costSoFar + wt < costings[nei]) {
                        costings[nei] = costSoFar + wt;
                        pq.offer(new Pair<>(costings[nei], nei));
                    }
                }
            }
        }

        if (visited.size() != n) {
            return -1;
        }

        for (int i = 1; i <= n; i++) {
            maxDelay = Math.max(maxDelay, costings[i]);
        }

        return maxDelay;
    }

    // Helper class for pairs
    static class Pair<K, V> {
        private K key;
        private V value;

        public Pair(K key, V value) {
            this.key = key;
            this.value = value;
        }

        public K getKey() {
            return key;
        }

        public V getValue() {
            return value;
        }
    }
}

File: 744-network-delay-time/network-delay-time.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        def build_graph(times: List[List[int]]) -> Dict[int, List[int]]:
            """builds the graph and return adjacency dictionary"""
            graph = defaultdict(list)
            for u, v, w in times: graph[u].append((v,w))
            return graph

        graph = build_graph(times)
        min_times = {}
        pq = [(0, k)] # cost, dist

        while pq:
            dcost, node = heapq.heappop(pq)
            if node in min_times: continue

            min_times[node] = dcost
            for neighbor, cost in graph[node]:
                if neighbor not in min_times or dcost + cost < min_times[node]:
                    heapq.heappush(pq, (cost+dcost, neighbor))

        return max(min_times.values()) if len(min_times) == n else -1